pyppeteer/lifecycle_watcher.py

```diff
@@ -2,175 +2,199 @@
+        else:
+            self._timeout_timer = self._loop.create_future()
+        return self._maximumTimer
+
+    def _navigatedWithinDocument(self, frame: Frame = None) -> None:
+        # note: frame never appears to specified, left in for compatibility
+        if frame == self._frame:
+            self._hasSameDocumentNavigation = True
+            self._checkLifecycleComplete()
+
+    def _checkLifecycleComplete(self) -> None:
+        if not self._checkLifecycle(self._frame, self._expectedLifecycle):
             return
-        self._hasSameDocumentNavigation = True
-        self._checkLifecycleComplete()
-
-    def _checkLifecycleComplete(self):
-        def check_lifecycle(frame: 'Frame', expected_lifecycle: List[str]):
-            for event in expected_lifecycle:
-                if event in frame._lifecycleEvents:
-                    return False
-            for child in frame.childFrames:
-                if not check_lifecycle(child, expected_lifecycle):
-                    return False
-            return True
-
-        if not check_lifecycle(self._frame, self._expectedLifecycle):
-            return
-        self._lifecycleCallback()
-
-        if self._frame._loaderId == self._initialLoaderId \
-                and not self._hasSameDocumentNavigation:
+        self._lifecycleFuture.set_result(None)
+        if self._frame._loaderId == self._initialLoaderId and not self._hasSameDocumentNavigation:
             return
+        if self._hasSameDocumentNavigation:
+            self._sameDocumentNavigationFuture.set_result(None)
         if self._frame._loaderId != self._initialLoaderId:
-            self._newDocumentNavigationCompleteCallback()
-
-    def dispose(self):
-        helper.removeEventListeners(self._eventListeners)
-        # js: clearTimeout(this._maximumTimer);  # todo, reimplement this?
-
-    def _sameDocumentNavigationCompleteCallback(self, value=None) -> None:
-        self._sameDocumentNavigationPromise.set_value(value)
-
-    def _lifecycleCallback(self, value=None) -> None:
-        self._lifecyclePromise.set_value(value)
-
-    def _newDocumentNavigationCompleteCallback(self, value=None) -> None:
-        self._lifecyclePromise.set_value(value)
-
-    def _terminationCallback(self, value=None) -> None:
-        self._lifecyclePromise.set_value(value)
+            self._newDocumentNavigationFuture.set_result(None)
 
+    def _checkLifecycle(self, frame: Frame, expectedLifecycle: List[str]) -> bool:
+        for event in expectedLifecycle:
+            if event not in frame._lifecycleEvents:
+                return False
+        for child in frame.childFrames:
+            if not self._checkLifecycle(child, expectedLifecycle):
+                return False
+        return True
 
+    def dispose(self) -> None:
+        self._cleanup()
 
+    def _cleanup(self) -> None:
+        helper.removeEventListeners(self._eventListeners)
+        # not sure if every future has to be canceled manually or not
```

Comment:
Again, not sure if it's really best practice here to cancel every future we've created or simply leave them hanging. Before this gets merged I'd like to resolve this comment

Reply:
As far as I understand the `dispose` method is only called for cleanup/descruction so it would be fair to clean up all of the hanging futures as well. I'd vote for cancel everything within try/except if necessary.

Reply:
Will implement. Can't hurt.

Reply:
I don't think a try/except inside would be necessary, the docs don't state anything about it ever raising: https://docs.python.org/3/library/asyncio-future.html#asyncio.Future.cancel